diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 32f48ef..a2e3125 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -10,7 +10,9 @@ endif()
 file(GLOB CUDA_SOURCES *.cu)
 file(GLOB CPP_SOURCES *.cpp)
 
-add_library(nvcomp SHARED ${CUDA_SOURCES} ${CPP_SOURCES})
+
+add_library(nvcomp STATIC ${CUDA_SOURCES} ${CPP_SOURCES})
+set_property(TARGET nvcomp PROPERTY POSITION_INDEPENDENT_CODE True)
 set_property(TARGET nvcomp PROPERTY CUDA_ARCHITECTURES ${GPU_ARCHS})
 target_compile_options(nvcomp PRIVATE
     $<$<COMPILE_LANGUAGE:CUDA>:--expt-extended-lambda -Xcompiler -pthread>)
