diff --git a/cpp/cmake_modules/FindRapidJSONAlt.cmake b/cpp/cmake_modules/FindRapidJSONAlt.cmake
index babb450e20..148dd93a78 100644
--- a/cpp/cmake_modules/FindRapidJSONAlt.cmake
+++ b/cpp/cmake_modules/FindRapidJSONAlt.cmake
@@ -26,7 +26,10 @@ endif()
 if(RapidJSONAlt_FIND_QUIETLY)
   list(APPEND find_package_args QUIET)
 endif()
+set(_CMAKE_POLICY_VERSION_MINIMUM_OLD ${CMAKE_POLICY_VERSION_MINIMUM})
+set(CMAKE_POLICY_VERSION_MINIMUM 3.5)
 find_package(RapidJSON ${find_package_args})
+set(CMAKE_POLICY_VERSION_MINIMUM ${_CMAKE_POLICY_VERSION_MINIMUM_OLD})
 if(RapidJSON_FOUND)
   set(RapidJSONAlt_FOUND TRUE)
   if(NOT TARGET RapidJSON)
