diff --git a/libcudacxx/include/cuda/std/detail/libcxx/include/__config b/libcudacxx/include/cuda/std/detail/libcxx/include/__config
index e7c62c031b..5db861853a 100644
--- a/libcudacxx/include/cuda/std/detail/libcxx/include/__config
+++ b/libcudacxx/include/cuda/std/detail/libcxx/include/__config
@@ -1049,7 +1049,6 @@ typedef __char32_t char32_t;
 #      define _LIBCUDACXX_EXPORTED_FROM_ABI __declspec(dllimport)
 #    endif
 
-#    define _LIBCUDACXX_TYPE_VIS      _LIBCUDACXX_DLL_VIS
 #    define _LIBCUDACXX_FUNC_VIS      _LIBCUDACXX_DLL_VIS
 #    define _LIBCUDACXX_EXCEPTION_ABI _LIBCUDACXX_DLL_VIS
 #    define _LIBCUDACXX_HIDDEN
@@ -1448,14 +1447,6 @@ __sanitizer_annotate_contiguous_container(const void*, const void*, const void*,
 #    define _LIBCUDACXX_WEAK __attribute__((__weak__))
 #  endif
 
-// Redefine some macros for internal use
-#  if defined(__cuda_std__)
-#    undef _LIBCUDACXX_FUNC_VIS
-#    define _LIBCUDACXX_FUNC_VIS _LIBCUDACXX_INLINE_VISIBILITY
-#    undef _LIBCUDACXX_TYPE_VIS
-#    define _LIBCUDACXX_TYPE_VIS
-#  endif // __cuda_std__
-
 // Thread API
 #  ifndef _LIBCUDACXX_HAS_THREAD_API_EXTERNAL
 #    if defined(_CCCL_COMPILER_NVRTC) || defined(__EMSCRIPTEN__)
