diff --git a/jitify.hpp b/jitify.hpp
index b38b48d..7914236 100644
--- a/jitify.hpp
+++ b/jitify.hpp
@@ -1220,7 +1220,7 @@ class CUDAKernel {
       if (_ptx[end] == '=') --end;
       std::string line = _ptx.substr(pos, end - pos);
       pos = end;
-      size_t symbol_start = line.find_last_of(" ") + 1;
+      size_t symbol_start = line.find_last_of(' ') + 1;
       size_t symbol_end = line.find_last_of("[");
       std::string entry = line.substr(symbol_start, symbol_end - symbol_start);
       std::string key = detail::demangle_ptx_variable_name(entry.c_str());
diff --git a/jitify2.hpp b/jitify2.hpp
index ede9fbf..905594d 100644
--- a/jitify2.hpp
+++ b/jitify2.hpp
@@ -2802,7 +2802,7 @@ inline void find_lowered_global_variables(StringRef ptx,
     if (ptx[end] == '=') --end;
     StringRef line = ptx.substr(pos, end - pos);
     pos = end;
-    size_t symbol_start = line.find_last_of(" ") + 1;
+    size_t symbol_start = line.find_last_of(' ') + 1;
     size_t symbol_end = line.find_last_of("[");
     std::string entry(line.substr(symbol_start, symbol_end - symbol_start));
     std::string key = demangle_ptx_variable_name(entry.c_str());
