diff --git a/nvbench/main.cuh b/nvbench/main.cuh
index 0ba82d7..7ab02c1 100644
--- a/nvbench/main.cuh
+++ b/nvbench/main.cuh
@@ -54,6 +54,14 @@
 // clang-format on
 #endif

+#ifndef NVBENCH_ENVIRONMENT
+namespace nvbench {
+struct no_environment
+{};
+}
+#define NVBENCH_ENVIRONMENT nvbench::no_environment
+#endif
+
 #define NVBENCH_MAIN_PARSE(argc, argv)                                                             \
   nvbench::option_parser parser;                                                                   \
   parser.parse(argc, argv)
@@ -77,6 +85,7 @@
     printer.set_total_state_count(total_states);                                                   \
                                                                                                    \
     printer.set_completed_state_count(0);                                                          \
+    auto env_state = NVBENCH_ENVIRONMENT();                                                        \
     for (auto &bench_ptr : benchmarks)                                                             \
     {                                                                                              \
       bench_ptr->set_printer(printer);                                                             \
