diff --git a/dependencies/cub/cub/device/dispatch/dispatch_radix_sort.cuh b/dependencies/cub/cub/device/dispatch/dispatch_radix_sort.cuh
index 953b24b0..3a895323 100644
--- a/dependencies/cub/cub/device/dispatch/dispatch_radix_sort.cuh
+++ b/dependencies/cub/cub/device/dispatch/dispatch_radix_sort.cuh
@@ -738,7 +738,7 @@ struct DeviceRadixSortPolicy


     /// SM60 (GP100)
-    struct Policy600 : ChainedPolicy<600, Policy600, Policy500>
+    struct Policy600 : ChainedPolicy<600, Policy600, Policy600>
     {
         enum {
             PRIMARY_RADIX_BITS      = (sizeof(KeyT) > 1) ? 7 : 5,    // 6.9B 32b keys/s (Quadro P100)
diff --git a/dependencies/cub/cub/device/dispatch/dispatch_reduce.cuh b/dependencies/cub/cub/device/dispatch/dispatch_reduce.cuh
index de182a93..e2ba1f4b 100644
--- a/dependencies/cub/cub/device/dispatch/dispatch_reduce.cuh
+++ b/dependencies/cub/cub/device/dispatch/dispatch_reduce.cuh
@@ -423,7 +423,7 @@ struct DeviceReducePolicy
   };

   /// SM60
-  struct Policy600 : ChainedPolicy<600, Policy600, Policy350>
+  struct Policy600 : ChainedPolicy<600, Policy600, Policy600>
   {
     static constexpr int threads_per_block  = 256;
     static constexpr int items_per_thread   = 16;
diff --git a/dependencies/cub/cub/device/dispatch/dispatch_scan.cuh b/dependencies/cub/cub/device/dispatch/dispatch_scan.cuh
index 0df89b7c..6c499d87 100644
--- a/dependencies/cub/cub/device/dispatch/dispatch_scan.cuh
+++ b/dependencies/cub/cub/device/dispatch/dispatch_scan.cuh
@@ -240,7 +240,7 @@ struct DeviceScanPolicy
   };

   /// SM600
-  struct Policy600 : ChainedPolicy<600, Policy600, Policy520>
+  struct Policy600 : ChainedPolicy<600, Policy600, Policy600>
   {
     typedef AgentScanPolicy<128,
                             15, ///< Threads per block, items per thread
