diff --git a/thrust/thrust/system/cuda/detail/dispatch.h b/thrust/thrust/system/cuda/detail/dispatch.h
index 3d004aa55..71ce86bea 100644
--- a/thrust/thrust/system/cuda/detail/dispatch.h
+++ b/thrust/thrust/system/cuda/detail/dispatch.h
@@ -63,7 +63,7 @@
   _THRUST_INDEX_TYPE_DISPATCH_GUARD_UNDERFLOW(count1)                \
   _THRUST_INDEX_TYPE_DISPATCH_GUARD_UNDERFLOW(count2)
 
-#if defined(THRUST_FORCE_64_BIT_OFFSET_TYPE)
+#if 0
 //! @brief Always dispatches to 64 bit offset version of an algorithm
 #  define THRUST_INDEX_TYPE_DISPATCH(status, call, count, arguments) \
     _THRUST_INDEX_TYPE_DISPATCH_GUARD_UNDERFLOW(count)               \
@@ -89,7 +89,7 @@
     _THRUST_INDEX_TYPE_DISPATCH_GUARD_UNDERFLOW(count)                                     \
     _THRUST_INDEX_TYPE_DISPATCH(std::uint64_t, status, call_64, count, arguments)
 
-#elif defined(THRUST_FORCE_32_BIT_OFFSET_TYPE)
+#elif 1
 
 //! @brief Ensures that the size of the input does not overflow the offset type
 #  define _THRUST_INDEX_TYPE_DISPATCH_GUARD_OVERFLOW(index_type, count)                       \
