--- thrust/system/cuda/detail/core/util.h.orig	2020-08-26 15:06:59.650997131 -0500
+++ thrust/system/cuda/detail/core/util.h	2020-08-26 15:08:41.338387987 -0500
@@ -652,7 +652,10 @@
   }
 
 #define CUDA_CUB_RET_IF_FAIL(e) \
-  if (cub::Debug((e), __FILE__, __LINE__)) return e;
+  do {                          \
+    auto const error = (e);     \
+    if (cub::Debug(error, __FILE__, __LINE__)) return error; \
+  } while(0);
 
   // uninitialized
   // -------
